src/Streamly/Internal/Data/Stream/StreamD.hs

```diff
@@ -4034,3 +4041,65 @@ lastN n = Fold step initial done
         foldFunc i
             | i < n = RB.unsafeFoldRingM
             | otherwise = RB.unsafeFoldRingFullM
+
+------------------------------------------------------------------------------
+-- Time related
+------------------------------------------------------------------------------
+
+{-# INLINE secondsToTimeSpec #-}
+secondsToTimeSpec :: Double -> TimeSpec
+secondsToTimeSpec d = TimeSpec (a :: Int64) (x :: Int64)  
+    where
+        tenPow9 = 10^(9 :: Int)
+        (a, b) = properFraction d
+        x = round $ b * tenPow9
+
+-- XXX This does not fuse. Removing the if function results in fusion though.
+-- XXX This function will fuse with the use of fusion-plugin.
+-- XXX This function also fuses if compiled with -funfolding-use-threshold=500
+{-# INLINE_NORMAL takeByTime #-}
+takeByTime :: MonadIO m => Double -> Stream m a -> Stream m a
+takeByTime lim' (Stream step state) = Stream step' (state, Nothing)
+  where
+    lim = secondsToTimeSpec lim'
```

Comment:
We can convert lim' to nanoseconds and then use `fromInteger` to convert nanoseconds to TimeSpec.